### PR TITLE
Tighten Constraint Bounds from Variables

### PR DESCRIPTION
**Pull request by sjkale**
Added Isomorphic Transformation that tightens the bounds on a constraint based on the variable bounds.
Note: same as previous pull request that I had messed up. See below
https://github.com/Pyomo/pyomo/pull/246

### CODE REVIEW

File: pyomo/solvers/plugins/solvers/multi.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
+import logging
```
Comment (qtothec) [inline pyomo/solvers/plugins/solvers/multi.py]: I don't think this file belongs in this PR.

Comment (qtothec) [reply pyomo/solvers/plugins/solvers/multi.py]: Oh, I see that it was deleted in subsequent commit. My bad. Ignore this.

File: pyomo/core/plugins/transform/interval_tightener.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+from pyomo.core.base.constraint import Constraint
+from pyomo.core.kernel.numvalue import value
+from pyomo.core.plugins.transform.hierarchy import IsomorphicTransformation
+from pyomo.repn.canonical_repn import generate_canonical_repn
+from pyomo.util.plugin import alias
+import logging
+import textwrap
+logger=logging.getLogger('pyomo.core')
+class BoundsTightener(IsomorphicTransformation):
+    """Tightens bound on constraint, by iterating through each variable and
```
Comment (carldlaird) [inline pyomo/core/plugins/transform/interval_tightener.py]: In my opinion, BoundsTightener is too general a name for this transformation. Bounds tightening is used in many difference contexts, including tightening the bounds on variables. How about "TightenConstraintBoundsFromVars" or something similar?

File: pyomo/core/plugins/transform/interval_tightener.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        super(BoundsTightener, self).__init__()
+
+    def _apply_to(self,instance):
+
+        for constr in instance.component_data_objects(ctype = Constraint,
+                                                        active = True,
+                                                        descend_into=True):
+            if not constr.body.polynomial_degree() == 1:
+                continue
+
```
Comment (carldlaird) [inline pyomo/core/plugins/transform/interval_tightener.py]: Looking at the code above, it seems that this transformation only applies to linear constraints. This caveat should appear in the code documentation for the class.

**Comment by qtothec** [comment]
Fyi @sjkale, you can also change the title of the PR

File: pyomo/core/plugins/transform/constraint_tightener.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                    if repn.variables[i].ub:
+                        UB = UB + coef * value(repn.variables[i].ub)
+                    else:
+                        UB = float('Inf')
+                    if repn.variables[i].lb:
+                        LB = LB + coef * value(repn.variables[i].lb)
+                    else:
+                        LB = -float('Inf')
+                else:
+                    if repn.variables[i].ub:
```
Comment (michaelbynum) [inline pyomo/core/plugins/transform/constraint_tightener.py]: Should this be "if repn.variables[i].lb:"?

File: pyomo/core/plugins/transform/constraint_tightener.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                        UB = UB + coef * value(repn.variables[i].lb)
+                    else:
+                        LB = -float('Inf')
```
Comment (michaelbynum) [inline pyomo/core/plugins/transform/constraint_tightener.py]: Should this be "UB = float('Inf')"?

File: pyomo/core/plugins/transform/constraint_tightener.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                    if repn.variables[i].lb:
```
Comment (michaelbynum) [inline pyomo/core/plugins/transform/constraint_tightener.py]: if repn.variables[i].ub?

File: pyomo/core/plugins/transform/constraint_tightener.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                        LB = LB + coef * value(repn.variables[i].ub)
+                    else:
+                        UB = float('Inf')
```
Comment (michaelbynum) [inline pyomo/core/plugins/transform/constraint_tightener.py]: LB = -float('Inf')

File: pyomo/core/plugins/transform/constraint_tightener.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+                else:
+                    constr._upper = float(UB)
+            if(constr.has_lb() and value(constr.lower) < LB) or (not constr.has_lb()):
+                if LB == -float('Inf'):
+                    constr._lower = None
+                else:
+                    constr._lower = float(LB)
+
+            if UB < LB:
+                logger.error("Infeasible variable bounds")
```
Comment (michaelbynum) [inline pyomo/core/plugins/transform/constraint_tightener.py]: Should this also raise an error?

File: pyomo/core/tests/transform/test_constraint_tightener.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        #Check for no coefficients
+        m = ConcreteModel()
+        m.v1 = Var(initialize=7, bounds=(7,10))
+        m.v2 = Var(initialize=2, bounds=(2,5))
+        m.v3 = Var(initialize=6, bounds=(6,9))
+        m.v4 = Var(initialize=1, bounds=(1,1))
+        m.c1 = Constraint(expr=m.v1 >= m.v2 + m.v3 + m.v4)
+
+        TransformationFactory('core.tighten_constraint_from_vars').apply_to(m)
+        self.assertTrue(m.c1._upper == 0)
```
Comment (michaelbynum) [inline pyomo/core/tests/transform/test_constraint_tightener.py]: Does the body of c1 come out to be "m.v1 - m.v2 - m.v3 - m.v4"? If so, shouldn't m.c1.upper be 1?

File: pyomo/core/tests/transform/test_constraint_tightener.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        m.v3 = Var(initialize=6, bounds=(6,9))
+        m.v4 = Var(initialize=1, bounds=(1,1))
+        m.c1 = Constraint(expr=m.v1 <= 2 *  m.v2 + m.v3 + m.v4)
+
+        TransformationFactory('core.tighten_constraint_from_vars').apply_to(m)
+        self.assertTrue(m.c1._upper == -1)
+        self.assertTrue(not m.c1.has_lb())
+        del m
+
+
```
Comment (michaelbynum) [inline pyomo/core/tests/transform/test_constraint_tightener.py]: Need tests with variables without bounds and with negative coefficients in the constraint body.

**Comment by blnicho** [comment]
@sjkale any response to the comments in Michael's review?

File: pyomo/core/plugins/transform/constraint_tightener.py (hunk 4 above)
Comment (qtothec) [reply pyomo/core/plugins/transform/constraint_tightener.py]: No (and applies to comments later on), because coef is negative in these cases.

File: pyomo/core/tests/transform/test_constraint_tightener.py (hunk 9 above)
Comment (qtothec) [reply pyomo/core/tests/transform/test_constraint_tightener.py]: No, it should be zero. It's subtraction, not division.

**Comment by blnicho** [comment]
@qtothec is this ready to review again?

**Comment by qtothec** [comment]
@blnicho One more commit. Going to change the tests to be clearer.

**Comment by qtothec** [comment]
@blnicho Ok, now it's ready to look at.
